// vidstack/dist/dev/index.js
import { AUDIO_PROVIDER } from './chunks/chunk-M667F2K3.js';
import { HLS_PROVIDER } from './chunks/chunk-SMG35QNF.js';
import { VIDEO_PROVIDER } from './chunks/chunk-7MNSRR4E.js';
import './chunks/chunk-XKJJT7VD.js';
export { TimeDefinition } from './chunks/chunk-IOSG4OXX.js';
export { ToggleButtonDefinition } from './chunks/chunk-HG7NKFCN.js';
export { VolumeSliderDefinition } from './chunks/chunk-KMYMG35E.js';
export { OutletDefinition } from './chunks/chunk-AYBAF5H3.js';
export { PlayerDefinition, createTimeRanges, getTimeRangesEnd, getTimeRangesStart, mediaStore, softResetMediaStore } from './chunks/chunk-6ABUXMT4.js';
import './chunks/chunk-K2X6MWIX.js';
export { PosterDefinition } from './chunks/chunk-6HNX7GKC.js';
export { MediaRemoteControl } from './chunks/chunk-FL2I7HKN.js';
export { SeekButtonDefinition } from './chunks/chunk-DWMLKJMI.js';
export { SliderThumbnailDefinition } from './chunks/chunk-CK5PXZGW.js';
export { SliderValueDefinition } from './chunks/chunk-QO76LTVQ.js';
export { SliderVideoDefinition } from './chunks/chunk-XTQ23N4D.js';
export { SliderDefinition } from './chunks/chunk-BEZMPZ66.js';
export { TimeSliderDefinition } from './chunks/chunk-TUVQXOYI.js';
import './chunks/chunk-JELAJF2G.js';
import './chunks/chunk-W2VHWD2S.js';
import './chunks/chunk-Z3USF3KA.js';
import './chunks/chunk-LNPU55YF.js';
import './chunks/chunk-TW2KE7QB.js';
import './chunks/chunk-CVLY5S52.js';
export { sliderStore, sliderStoreContext, useSliderStore } from './chunks/chunk-J5ZUXZE6.js';
export { FullscreenButtonDefinition } from './chunks/chunk-4SRAJU4F.js';
export { MediaIconDefinition } from './chunks/chunk-GB67J7MB.js';
export { LiveIndicatorDefinition } from './chunks/chunk-O64PWQBX.js';
export { MuteButtonDefinition } from './chunks/chunk-B4FRYNOD.js';
export { PlayButtonDefinition } from './chunks/chunk-QAG4D3G2.js';
import './chunks/chunk-7L4PUHMW.js';
export { MEDIA_KEY_SHORTCUTS } from './chunks/chunk-HNFYBM6K.js';
import './chunks/chunk-GLI6O3AL.js';
import './chunks/chunk-25YO7G2G.js';
import './chunks/chunk-YQSJJLRL.js';
export { mediaContext } from './chunks/chunk-3ULVZKKX.js';
export { appendTriggerEvent, findTriggerEvent, hasTriggerEvent, walkTriggerEventChain } from 'maverick.js/std';

// src/player/media/providers/type-check.ts
function isAudioProvider(provider) {
  return !!provider?.[AUDIO_PROVIDER];
}
function isVideoProvider(provider) {
  return !!provider?.[VIDEO_PROVIDER];
}
function isHLSProvider(provider) {
  return !!provider?.[HLS_PROVIDER];
}
function isHTMLAudioElement(element) {
  return element instanceof HTMLAudioElement;
}
function isHTMLVideoElement(element) {
  return element instanceof HTMLVideoElement;
}

// src/index.ts
{
  console.warn("[vidstack] dev mode!");
}

export { isAudioProvider, isHLSProvider, isHTMLAudioElement, isHTMLVideoElement, isVideoProvider };
