// maverick.js/dist/server/element.cjs
'use strict';

var chunkGUOGBKY3_cjs = require('./chunks/chunk-GUOGBKY3.cjs');
require('./chunks/chunk-BCLCMONJ.cjs');
var chunkCTEVPPZN_cjs = require('./chunks/chunk-CTEVPPZN.cjs');
var chunkH22IBEBC_cjs = require('./chunks/chunk-H22IBEBC.cjs');
require('./chunks/chunk-AH67QMAU.cjs');



Object.defineProperty(exports, 'ARRAY', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.ARRAY; }
});
Object.defineProperty(exports, 'BOOLEAN', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.BOOLEAN; }
});
Object.defineProperty(exports, 'FUNCTION', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.FUNCTION; }
});
Object.defineProperty(exports, 'NUMBER', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.NUMBER; }
});
Object.defineProperty(exports, 'OBJECT', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.OBJECT; }
});
Object.defineProperty(exports, 'STRING', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.STRING; }
});
Object.defineProperty(exports, 'adoptCSS', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.adoptCSS; }
});
Object.defineProperty(exports, 'createHTMLElement', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.createHTMLElement; }
});
Object.defineProperty(exports, 'createServerElement', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.createServerElement; }
});
Object.defineProperty(exports, 'css', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.css; }
});
Object.defineProperty(exports, 'defineCustomElement', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.defineCustomElement; }
});
Object.defineProperty(exports, 'inferAttributeType', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.inferAttributeType; }
});
Object.defineProperty(exports, 'injectCSS', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.injectCSS; }
});
Object.defineProperty(exports, 'onAttach', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.onAttach; }
});
Object.defineProperty(exports, 'onConnect', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.onConnect; }
});
Object.defineProperty(exports, 'registerCustomElement', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.registerCustomElement; }
});
Object.defineProperty(exports, 'registerHeadlessCustomElement', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.registerHeadlessCustomElement; }
});
Object.defineProperty(exports, 'registerLiteCustomElement', {
  enumerable: true,
  get: function () { return chunkGUOGBKY3_cjs.registerLiteCustomElement; }
});
Object.defineProperty(exports, 'createElementInstance', {
  enumerable: true,
  get: function () { return chunkCTEVPPZN_cjs.createElementInstance; }
});
Object.defineProperty(exports, 'MEMBERS', {
  enumerable: true,
  get: function () { return chunkH22IBEBC_cjs.MEMBERS; }
});
Object.defineProperty(exports, 'PROPS', {
  enumerable: true,
  get: function () { return chunkH22IBEBC_cjs.PROPS; }
});
Object.defineProperty(exports, 'RENDER', {
  enumerable: true,
  get: function () { return chunkH22IBEBC_cjs.RENDER; }
});
Object.defineProperty(exports, 'getCustomElementInstance', {
  enumerable: true,
  get: function () { return chunkH22IBEBC_cjs.getCustomElementInstance; }
});
