// @maverick-js/signals/dist/server/index.cjs
'use strict';

var chunkQJ2T6T4S_cjs = require('./chunks/chunk-QJ2T6T4S.cjs');



Object.defineProperty(exports, 'SCOPE', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.SCOPE; }
});
Object.defineProperty(exports, 'computed', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.computed; }
});
Object.defineProperty(exports, 'createComputation', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.createComputation; }
});
Object.defineProperty(exports, 'createScope', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.createScope; }
});
Object.defineProperty(exports, 'effect', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.effect; }
});
Object.defineProperty(exports, 'getContext', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.getContext; }
});
Object.defineProperty(exports, 'getScope', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.getScope; }
});
Object.defineProperty(exports, 'isFunction', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.isFunction; }
});
Object.defineProperty(exports, 'isNotEqual', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.isNotEqual; }
});
Object.defineProperty(exports, 'isReadSignal', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.isReadSignal; }
});
Object.defineProperty(exports, 'isWriteSignal', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.isWriteSignal; }
});
Object.defineProperty(exports, 'onDispose', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.onDispose; }
});
Object.defineProperty(exports, 'onError', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.onError; }
});
Object.defineProperty(exports, 'peek', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.peek; }
});
Object.defineProperty(exports, 'readonly', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.readonly; }
});
Object.defineProperty(exports, 'root', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.root; }
});
Object.defineProperty(exports, 'scoped', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.scoped; }
});
Object.defineProperty(exports, 'setContext', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.setContext; }
});
Object.defineProperty(exports, 'signal', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.signal; }
});
Object.defineProperty(exports, 'tick', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.tick; }
});
Object.defineProperty(exports, 'untrack', {
  enumerable: true,
  get: function () { return chunkQJ2T6T4S_cjs.untrack; }
});
