// vidstack/dist/server/index.js
import { AUDIO_PROVIDER } from './chunks/chunk-MGDG6I4V.js';
import { HLS_PROVIDER } from './chunks/chunk-4N2AZASS.js';
import { VIDEO_PROVIDER } from './chunks/chunk-U654IJ5Z.js';
import './chunks/chunk-R25FJ2GA.js';
export { TimeDefinition } from './chunks/chunk-WQ773SEZ.js';
export { ToggleButtonDefinition } from './chunks/chunk-TRAUVZ55.js';
export { VolumeSliderDefinition } from './chunks/chunk-FWIPARRA.js';
export { OutletDefinition } from './chunks/chunk-KXV352XC.js';
export { PlayerDefinition, createTimeRanges, getTimeRangesEnd, getTimeRangesStart, mediaStore, softResetMediaStore } from './chunks/chunk-X7B54CIG.js';
import './chunks/chunk-K63O6LL7.js';
export { PosterDefinition } from './chunks/chunk-LTTFM5GQ.js';
export { MediaRemoteControl } from './chunks/chunk-PU7R6IPU.js';
export { SeekButtonDefinition } from './chunks/chunk-DTFSCEI7.js';
export { SliderThumbnailDefinition } from './chunks/chunk-UNCSPTKH.js';
export { SliderValueDefinition } from './chunks/chunk-MN3MKFGI.js';
export { SliderVideoDefinition } from './chunks/chunk-KUBBE23R.js';
export { SliderDefinition } from './chunks/chunk-43DIHY47.js';
export { TimeSliderDefinition } from './chunks/chunk-PLSH3G7K.js';
import './chunks/chunk-JELAJF2G.js';
import './chunks/chunk-W2VHWD2S.js';
import './chunks/chunk-QYXIDTPL.js';
import './chunks/chunk-OJUUTMEH.js';
import './chunks/chunk-TW2KE7QB.js';
import './chunks/chunk-N2X5VJTG.js';
export { sliderStore, sliderStoreContext, useSliderStore } from './chunks/chunk-J5ZUXZE6.js';
export { FullscreenButtonDefinition } from './chunks/chunk-EUQX3P2B.js';
export { MediaIconDefinition } from './chunks/chunk-CD65GM7M.js';
export { LiveIndicatorDefinition } from './chunks/chunk-WXCRUBAT.js';
export { MuteButtonDefinition } from './chunks/chunk-66WRP7SJ.js';
export { PlayButtonDefinition } from './chunks/chunk-LCR36PCE.js';
import './chunks/chunk-WQ5GROIP.js';
export { MEDIA_KEY_SHORTCUTS } from './chunks/chunk-HHVUXD4K.js';
import './chunks/chunk-IWORFIII.js';
import './chunks/chunk-EWIZ7YX3.js';
import './chunks/chunk-FU2MDXXS.js';
export { mediaContext } from './chunks/chunk-3ULVZKKX.js';
export { appendTriggerEvent, findTriggerEvent, hasTriggerEvent, walkTriggerEventChain } from 'maverick.js/std';

// src/player/media/providers/type-check.ts
function isAudioProvider(provider) {
  return !!(provider == null ? void 0 : provider[AUDIO_PROVIDER]);
}
function isVideoProvider(provider) {
  return !!(provider == null ? void 0 : provider[VIDEO_PROVIDER]);
}
function isHLSProvider(provider) {
  return !!(provider == null ? void 0 : provider[HLS_PROVIDER]);
}
function isHTMLAudioElement(element) {
  return false;
}
function isHTMLVideoElement(element) {
  return false;
}

export { isAudioProvider, isHLSProvider, isHTMLAudioElement, isHTMLVideoElement, isVideoProvider };
