// maverick.js/dist/server/chunks/chunk-CTEVPPZN.cjs
'use strict';

var chunkH22IBEBC_cjs = require('./chunk-H22IBEBC.cjs');
var chunkAH67QMAU_cjs = require('./chunk-AH67QMAU.cjs');
var signals = require('@maverick-js/signals');

function createElementInstance(definition, init = {}) {
  return signals.root((dispose) => {
    var _a;
    if (init.scope)
      init.scope.append(signals.getScope());
    let accessors = null, destroyed = false, hasProps = "props" in definition, $props = hasProps ? createInstanceProps(definition.props) : {}, $connected = signals.signal(false), $attrs = {}, $styles = {}, setAttributes = (attrs) => void Object.assign($attrs, attrs), setStyles = (styles) => void Object.assign($styles, styles);
    if (init.props && hasProps) {
      for (const prop of Object.keys(init.props)) {
        if (prop in definition.props)
          $props["$" + prop].set(init.props[prop]);
      }
    }
    const host = {
      [chunkH22IBEBC_cjs.PROPS]: {
        $attrs,
        $styles,
        $connected
      },
      el: null,
      $el() {
        return $connected() ? host.el : null;
      },
      $connected,
      setAttributes,
      setStyles,
      setCSSVars: setStyles
    };
    const instance = {
      host,
      props: $props,
      [chunkH22IBEBC_cjs.SCOPE]: signals.getScope(),
      [chunkH22IBEBC_cjs.PROPS]: $props,
      [chunkH22IBEBC_cjs.ATTACH]: [],
      [chunkH22IBEBC_cjs.CONNECT]: [],
      accessors() {
        if (accessors)
          return accessors;
        const props = {};
        for (const prop of Object.keys(definition.props))
          props[prop] = $props["$" + prop];
        return accessors = chunkAH67QMAU_cjs.createAccessors(props);
      },
      destroy() {
        var _a2;
        if (destroyed)
          return;
        destroyed = true;
        (_a2 = host.el) == null ? void 0 : _a2.destroy();
        instance[chunkH22IBEBC_cjs.ATTACH].length = 0;
        instance[chunkH22IBEBC_cjs.CONNECT].length = 0;
        dispose();
        instance[chunkH22IBEBC_cjs.SCOPE] = null;
        instance[chunkH22IBEBC_cjs.MEMBERS] = null;
        instance[chunkH22IBEBC_cjs.RENDER] = null;
        host.el = null;
      }
    };
    try {
      chunkH22IBEBC_cjs.setCustomElementInstance(instance);
      instance[chunkH22IBEBC_cjs.MEMBERS] = definition.setup(instance);
    } finally {
      chunkH22IBEBC_cjs.setCustomElementInstance(null);
    }
    const $render = (_a = instance[chunkH22IBEBC_cjs.MEMBERS]) == null ? void 0 : _a.$render;
    if ($render) {
      instance[chunkH22IBEBC_cjs.RENDER] = function render() {
        let result = null;
        signals.scoped(() => {
          try {
            chunkH22IBEBC_cjs.setCustomElementInstance(instance);
            result = $render();
          } finally {
            chunkH22IBEBC_cjs.setCustomElementInstance(null);
          }
        }, instance[chunkH22IBEBC_cjs.SCOPE]);
        return result;
      };
    }
    signals.onDispose(instance.destroy);
    return instance;
  });
}
function createInstanceProps(propDefs) {
  const props = {};
  for (const name of Object.keys(propDefs)) {
    const def = propDefs[name];
    props["$" + name] = signals.signal(def.initial, def);
  }
  return props;
}

exports.createElementInstance = createElementInstance;
