// @maverick-js/signals/dist/prod/map.js
import { effect, write, onDispose, read, isNotEqual, createComputation, createScope, scoped, dispose, compute } from './chunks/chunk-JI2WIKTK.js';

// src/selector.ts
function selector(source) {
  let currentKey, nodes = /* @__PURE__ */ new Map();
  effect(() => {
    const newKey = source(), prev = nodes.get(currentKey), next = nodes.get(newKey);
    prev && write.call(prev, false);
    next && write.call(next, true);
    currentKey = newKey;
  });
  return function observeSelector(key) {
    let node = nodes.get(key);
    if (!node)
      nodes.set(key, node = new Selector(key, key === currentKey, nodes));
    node.r += 1;
    onDispose(node);
    return read.bind(node);
  };
}
function Selector(key, initialValue, nodes) {
  this.f = 0;
  this.u = key;
  this.j = initialValue;
  this.r = 0;
  this.c = nodes;
  this.d = null;
}
var SelectorProto = Selector.prototype;
SelectorProto.p = isNotEqual;
SelectorProto.call = function() {
  this.r -= 1;
  if (!this.r) {
    this.c.delete(this.u);
    this.c = null;
  }
};

// src/map.ts
function computedMap(list, map, options) {
  return read.bind(
    createComputation(
      [],
      updateMap.bind({
        m: createScope(),
        h: 0,
        s: list,
        g: [],
        n: map,
        e: [],
        c: []
      }),
      options
    )
  );
}
function updateMap() {
  let i = 0, newItems = this.s() || [], mapper = () => this.n(read.bind(this.c[i]), i);
  scoped(() => {
    if (newItems.length === 0) {
      if (this.h !== 0) {
        dispose.call(this.m, false);
        this.g = [];
        this.e = [];
        this.h = 0;
        this.c = [];
      }
      return;
    }
    for (i = 0; i < newItems.length; i++) {
      if (i < this.g.length && this.g[i] !== newItems[i]) {
        write.call(this.c[i], newItems[i]);
      } else if (i >= this.g.length) {
        this.e[i] = compute(
          this.c[i] = createComputation(newItems[i], null),
          mapper,
          null
        );
      }
    }
    for (; i < this.g.length; i++)
      dispose.call(this.c[i]);
    this.h = this.c.length = newItems.length;
    this.g = newItems.slice(0);
    this.e = this.e.slice(0, this.h);
  }, this.m);
  return this.e;
}
function computedKeyedMap(list, map, options) {
  return read.bind(
    createComputation(
      [],
      updateKeyedMap.bind({
        m: createScope(),
        h: 0,
        s: list,
        g: [],
        n: map,
        e: [],
        c: []
      }),
      options
    )
  );
}
function updateKeyedMap() {
  const newItems = this.s() || [], indexed = this.n.length > 1;
  scoped(() => {
    let newLen = newItems.length, i, j, mapper = indexed ? () => this.n(newItems[j], read.bind(this.c[j])) : () => this.n(newItems[j]);
    if (newLen === 0) {
      if (this.h !== 0) {
        dispose.call(this.m, false);
        this.c = [];
        this.g = [];
        this.e = [];
        this.h = 0;
      }
    } else if (this.h === 0) {
      this.e = new Array(newLen);
      for (j = 0; j < newLen; j++) {
        this.g[j] = newItems[j];
        this.e[j] = compute(
          this.c[j] = createComputation(j, null),
          mapper,
          null
        );
      }
      this.h = newLen;
    } else {
      let start, end, newEnd, item, newIndices, newIndicesNext, temp = new Array(newLen), tempNodes = new Array(newLen);
      for (start = 0, end = Math.min(this.h, newLen); start < end && this.g[start] === newItems[start]; start++)
        ;
      for (end = this.h - 1, newEnd = newLen - 1; end >= start && newEnd >= start && this.g[end] === newItems[newEnd]; end--, newEnd--) {
        temp[newEnd] = this.e[end];
        tempNodes[newEnd] = this.c[end];
      }
      newIndices = /* @__PURE__ */ new Map();
      newIndicesNext = new Array(newEnd + 1);
      for (j = newEnd; j >= start; j--) {
        item = newItems[j];
        i = newIndices.get(item);
        newIndicesNext[j] = i === void 0 ? -1 : i;
        newIndices.set(item, j);
      }
      for (i = start; i <= end; i++) {
        item = this.g[i];
        j = newIndices.get(item);
        if (j !== void 0 && j !== -1) {
          temp[j] = this.e[i];
          tempNodes[j] = this.c[i];
          j = newIndicesNext[j];
          newIndices.set(item, j);
        } else
          dispose.call(this.c[i]);
      }
      for (j = start; j < newLen; j++) {
        if (j in temp) {
          this.e[j] = temp[j];
          this.c[j] = tempNodes[j];
          write.call(this.c[j], j);
        } else {
          this.e[j] = compute(
            this.c[j] = createComputation(j, null),
            mapper,
            null
          );
        }
      }
      this.e = this.e.slice(0, this.h = newLen);
      this.g = newItems.slice(0);
    }
  }, this.m);
  return this.e;
}

export { computedKeyedMap, computedMap, selector };
